### DISCUSSION TITLE
CLI is getting car data, python script says: invalid_client

### DISCUSSION THREAD

**Discussion by cloudswitch** [bimmerconnected/bimmer_connected#704] [General] [answered]
Hi, sorry if this is a duplicate. I'm new to Python (speak PHP a little bit though) and maybe it's a lame question so please excuse me

I'm on Raspberry Pi 5 with Raspberry Pi OS.
I have installed python 3.11.2, created and activated a virtual environment, installed bimmer_connected (0.17.2).

when I type in the CLI:
`bimmerconnected status USERNAME "PASSWORD" rest_of_world --captcha-token "TOKEN"`
Then I'm getting the car data.

But when using the example py script with exactly the same credentials, I'm getting:

> bimmer_connected.models.MyBMWAuthError: HTTPStatusError: invalid_client - Client authentication failed (e.g., login failure, unknown client, no client authentication included or unsupported authentication method)

What could be the reason for that?

I can sign in in my bmw app and on the web version without any problems.

Thank you.

**Comment by rikroe** [comment]
Are you using a new captcha token for the Python variant?

**Reply by cloudswitch** [reply]
new? Is there a new and/or old one as well?
I was using the one that's in the docs.
https://bimmer-connected.readthedocs.io/en/stable/captcha/rest_of_world.html

**Reply by rikroe** [reply]
Check the warning in https://bimmer-connected.readthedocs.io/en/stable/captcha.html ;)
The captcha tokens are one-time use only, you need to generate a "new" one everytime you want to log in (not on the CLI, but thats the additional development that is needed).

**Reply by cloudswitch** [reply]
But if we need a new token every time for the py script to work, than it's impossible to automate the process, isn't it?

**Reply by rikroe** [reply]
https://bimmer-connected.readthedocs.io/en/stable/captcha.html#using-the-python-api see the last very last paragraph.

You need to build something similar to the CLI's `load_oauth_store_from_file()` and `store_oauth_store_to_file()` functions, depending on how you are using it and where you are importing it.
You'll need at least `refresh_token` and `gcid`.

The Home Assistant component does store the information within the HA entity registry. There is also another implementation here: https://github.com/FilipDem/Domoticz-BMW-plugin.

**Reply by cloudswitch** [reply]
one more question:

> For storing the data across restarts, an example implementation can be found in bimmerconnected.cli.main() with load_oauth_store_from_file() and store_oauth_store_to_file().

where do I find these?

**Reply by rikroe** [reply]
https://github.com/bimmerconnected/bimmer_connected/blob/master/bimmer_connected/cli.py#L328-L363

**Reply by cloudswitch** [reply]
Okay, I would simply like to **automatically** get the SOC once a day, calculate the charging amount needed to reach 80%, calculate the starting time back from the target time (6:15 in the morning), and send it to the charger API. Basically it should be not that complicated, but still...

**Reply by cloudswitch** [reply]
> https://github.com/bimmerconnected/bimmer_connected/blob/master/bimmer_connected/cli.py#L328-L363

thank you!

**Reply by rikroe** [reply, marked as answer]
There are a couple of options:
- try using the Python API and develop the script yourself (probably the hardest)
- use Python/NodeRED or any other automation tool you have experience with to call the CLI with the `--json` argument `bimmerconnected status --json USERNAME PASSWORD REGION` to only get a JSON output that can be parsed in a third party tool
- Try Home Assistant, use the integrated BMW component and build your own automations for this case

**Reply by cloudswitch** [reply]
okay I'm going to give it a try. is it possible to get `vehicle.fuel_and_battery.remaining_battery_percent` only? (I was trying to submit the VIN in the CLI as well, but got always data for both cars.)

**Reply by rikroe** [reply]
You should be able to get data for one car only by using `bimmerconnected status --json -v VIN USERNAME PASSWORD REGION`.
It will give all details of the car. You could try parsing the value with `jq` to get a single value (there's loads of tutorials for this one).

**Reply by cloudswitch** [reply]
> You should be able to get data for one car only by using `bimmerconnected status --json -v VIN USERNAME PASSWORD REGION`. It will give all details of the car. You could try parsing the value with `jq` to get a single value (there's loads of tutorials for this one).

if I use `--json` it's working, without not. 👍

**Comment by cloudswitch** [comment]
I have the script and it's working. I just have to add it to crontab 👍 
many thanks for your help
